emulator/app/src/mctp_transport.rs

```diff
@@ -84,26 +95,47 @@ impl PldmSocket for MctpPldmSocket {
     }
 
     fn receive(&self, _timeout: Option<Duration>) -> Result<RxPacket, PldmTransportError> {
-        let context = &mut *self.context.lock().unwrap();
-        if let Some(response) = context.first_response.as_mut() {
-            let mut data = [0u8; MAX_PLDM_PAYLOAD_SIZE];
-            // Skip the first byte containing the MCTP common header
-            // and only return the PLDM payload
-            data[..response.len() - 1].copy_from_slice(&response[1..]);
-            let ret = RxPacket {
-                src: self.dest,
-                payload: Payload {
-                    data,
-                    len: response.len() - 1,
-                },
-            };
-            context.first_response = None;
-            return Ok(ret);
+        {
+            let (context_lock, cvar) = &*self.context;
+            let mut context = context_lock.lock().unwrap();
+            if context.state == MctpPldmSocketState::FirstResponse
+                || context.state == MctpPldmSocketState::Idle
+            {
+                if context.first_response.is_none() {
```

Comment:
When using condition variables, you need to check the condition after you wake up. Although it is unlikely here, there are other reasons the thread could have woken up.

```suggestion
                while context.first_response.is_none() {
```

Reply:
thanks! that's a good point.